// tailwindcss-patch/dist/index.cjs
'use strict';

const index = require('./shared/tailwindcss-patch.2fb4223c.cjs');
const config = require('@tailwindcss-mangle/config');
require('node:path');
require('node:fs');
require('node:process');
require('semver');
require('@babel/types');
require('@babel/generator');
require('@babel/traverse');
require('@babel/parser');
require('node:fs/promises');
require('resolve');
require('postcss');
require('lilconfig');
require('jiti');



exports.CacheManager = index.CacheManager;
exports.TailwindcssPatcher = index.TailwindcssPatcher;
exports.ensureDir = index.ensureDir;
exports.ensureFileContent = index.ensureFileContent;
exports.getCacheOptions = index.getCacheOptions;
exports.getPackageInfoSync = index.getPackageInfoSync;
exports.inspectPostcssPlugin = index.inspectPostcssPlugin;
exports.inspectProcessTailwindFeaturesReturnContext = index.inspectProcessTailwindFeaturesReturnContext;
exports.internalPatch = index.internalPatch;
exports.isObject = index.isObject;
exports.monkeyPatchForExposingContextV2 = index.monkeyPatchForExposingContextV2;
exports.monkeyPatchForExposingContextV3 = index.monkeyPatchForExposingContextV3;
exports.requireResolve = index.requireResolve;
exports.defineConfig = config.defineConfig;
