// unplugin-tailwindcss-mangle/dist/shared/unplugin-tailwindcss-mangle.43bd55b2.cjs
'use strict';

const fs = require('node:fs/promises');
const path = require('node:path');
const shared = require('@tailwindcss-mangle/shared');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
const path__default = /*#__PURE__*/_interopDefaultCompat(path);

const pluginName = "unplugin-tailwindcss-mangle";

function escapeStringRegexp(str) {
  if (typeof str !== "string") {
    throw new TypeError("Expected a string");
  }
  return str.replaceAll(/[$()*+.?[\\\]^{|}]/g, "\\$&").replaceAll("-", "\\x2d");
}
function getGroupedEntries(entries, options = {
  cssMatcher(file) {
    return /\.css$/.test(file);
  },
  htmlMatcher(file) {
    return /\.html?$/.test(file);
  },
  jsMatcher(file) {
    return /\.[cm]?js$/.test(file);
  }
}) {
  const { cssMatcher, htmlMatcher, jsMatcher } = options;
  const groupedEntries = shared.groupBy(entries, ([file]) => {
    if (cssMatcher(file)) {
      return "css";
    } else if (htmlMatcher(file)) {
      return "html";
    } else if (jsMatcher(file)) {
      return "js";
    } else {
      return "other";
    }
  });
  if (!groupedEntries.css) {
    groupedEntries.css = [];
  }
  if (!groupedEntries.html) {
    groupedEntries.html = [];
  }
  if (!groupedEntries.js) {
    groupedEntries.js = [];
  }
  if (!groupedEntries.other) {
    groupedEntries.other = [];
  }
  return groupedEntries;
}
function getCacheDir(basedir = process.cwd()) {
  return path__default.resolve(basedir, "node_modules/.cache", pluginName);
}
async function ensureDir(p) {
  try {
    await fs__default.access(p);
  } catch {
    await fs__default.mkdir(p, {
      recursive: true
    });
  }
}

exports.ensureDir = ensureDir;
exports.escapeStringRegexp = escapeStringRegexp;
exports.getCacheDir = getCacheDir;
exports.getGroupedEntries = getGroupedEntries;
exports.pluginName = pluginName;
